contracts/AdvancedProviders/MultiWithdrawProvider/MultiWithdrawProvider.sol

```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "./MultiWithdrawState.sol";
+
+contract MultiWithdrawProvider is MultiWithdrawState {
+    constructor(ILockDealNFT nftContract, uint256 _maxPoolsPerTx) {
+        name = "MultiWithdrawProvider";
+        lockDealNFT = nftContract;
+        maxPoolsPerTx = _maxPoolsPerTx;
+    }
+
+    function multiWithdrawAllPoolsOfOwner(address _owner) external {
+        uint256[] memory poolIds = getAllPoolsOfOwner(_owner);
+        _processMultiWithdraw(poolIds, _owner);
+    }
+
+    function multiWithdrawPoolsOfOwnerByVault(address _owner, uint256 _vaultId) external {
+        uint256[] memory poolIds = getPoolsOfOwnerByVault(_owner, _vaultId);
+        _processMultiWithdraw(poolIds, _owner);
+    }
+
+    function multiWithdrawPools(uint256[] memory _poolIds) external {
+        _processMultiWithdraw(_poolIds);
+    }
+
+    function getAllPoolsOfOwner(address _owner) public view returns (uint256[] memory) {
+        uint256 totalPools = lockDealNFT.balanceOf(_owner);
+        uint256[] memory poolIds = new uint256[](totalPools);
+        for(uint256 i = 0; i < totalPools; ) {
+            poolIds[i] = lockDealNFT.tokenOfOwnerByIndex(_owner, i);
+            unchecked { ++i; }
+        }
+        return poolIds;
+    }
+
+    function getPoolsOfOwnerByVault(address _owner, uint256 _vaultId) public view returns (uint256[] memory) {
+        uint256 totalPools = lockDealNFT.balanceOf(_owner);
+        uint256[] memory poolIds = new uint256[](totalPools);
+        uint256 index;
+        for(uint256 i = 0; i < totalPools; ) {
+            uint256 poolId = lockDealNFT.tokenOfOwnerByIndex(_owner, i);
+            if (lockDealNFT.poolIdToVaultId(poolId) == _vaultId) {
```

Comment:
we can use the `getData`.`vaultId`
https://github.com/The-Poolz/LockDealNFT/blob/e67315183004464639e4f00f2aa781250e55f606/contracts/interfaces/ILockDealNFT.sol#L30C17-L30C24